### PR TITLE
Drop tens of thousands of object allocations on require

### PR DESCRIPTION
**Pull request by schneems**
This patch utilizes a few techniques to cut down object allocations.and speed up requiring `mime/types`

- Use frozen strings and in place string manipulation
- Use self modifying operators when possible i.e. gsub!
- Avoiding assigning multiple variables using the splat operator (i.e. `a, b = *[1,2]`) as this allocates an extra un-needed array.

The call to `self.friendly({})` serves only to set the instance variable. The method call here merges an empty hash with an empty hash we can get rid of it.

To benchmark:

```
require 'allocation_tracer'

$LOAD_PATH << "/Users/schneems/Documents/projects/mime-types/lib"

ObjectSpace::AllocationTracer.trace do
  require 'mime/types'
end

puts :TOTAL => ObjectSpace::AllocationTracer.allocated_count_table.values.inject(:+)
```

Run without patch:

```
{:TOTAL=>195692}
```

Run with patch:

```
{:TOTAL=>158240}
```

### CODE REVIEW

File: lib/mime/type.rb [hunk 1]
```diff
@@ -618,8 +618,10 @@ def simplified(content_type)
 
       if matchdata
         matchdata.captures.map { |e|
-          e.downcase.gsub(UNREGISTERED_RE, '')
-        }.join('/')
+          e.downcase!
+          e.gsub!(UNREGISTERED_RE, ''.freeze)
```
Comment (halostatue) [inline lib/mime/type.rb]: Wouldn’t this be better as a constant?

File: lib/mime/type.rb [hunk 2]
```diff
@@ -742,10 +747,13 @@ def content_type=(type_string)
     raise InvalidContentType, type_string if match.nil?
 
     @content_type                  = type_string
-    @raw_media_type, @raw_sub_type = *match.captures
+    captures                       = match.captures
+    @raw_media_type                = captures.first
+    @raw_sub_type                  = captures.last
```
Comment (halostatue) [inline lib/mime/type.rb]: I was able to shave off a ridiculously small number of allocations (~9?) by changing this and the equivalent to:

```ruby
@raw_media_type, @raw_sub_type = match.captures
```

Nothing compared to what you’ve been able to reduce.

**Comment by halostatue** [comment]
This looks awesome, and I have a couple of comments/questions in-line. I’ve run benchmarks of this against `master`; this looks like it also brings 3–4% load performance boost.

File: lib/mime/type.rb (hunk 1 above)
Comment (schneems) [inline lib/mime/type.rb]: It's a wash, the constant requires an extra lookup.

Run this benchmark:

```ruby
require 'benchmark/ips'

FROZEN = "".freeze
Benchmark.ips do |x|
  x.report("constant") { "".gsub!(//, FROZEN) }
  x.report("frozen string") { "".gsub!(//, "".freeze) }
end     
```

And you'll see each side "win" often enough. Generally people put them in constants for aesthetic reasons or if they're being re-used all over the project. I didn't want to mess with aesthetics. If you want to push them to constants you can, but I would keep that a separate PR and discussion and keep this one focused mostly on perf.

As a side note, you don't need to constantize regexs, Ruby regexes that are static are effectively frozen by default, they only instantiate one object.

File: lib/mime/type.rb (hunk 2 above)
Comment (schneems) [inline lib/mime/type.rb]: :+1: I modified my PR, every little bit helps :grinning: I'll modify my PR and push it up.

**Comment by schneems** [comment]
I pushed up the changes you suggested. If you or anyone else is interested how I found these hotspots, I used this:


```ruby
require 'allocation_tracer'
require 'pretty_print'

$LOAD_PATH << "/Users/schneems/Documents/projects/mime-types/lib"

ObjectSpace::AllocationTracer.setup(%i{path line type})

r = ObjectSpace::AllocationTracer.trace do
  require 'mime/types'
end

pp r.sort {|x, y| x.last.first <=> y.last.first }.reverse.first(100)
```

You also get similar info with `memory_profiler` gem but the allocation tracer was a bit faster.

**Comment by halostatue** [comment]
Can I ask for one more change?

```diff
diff --git a/Rakefile b/Rakefile
index b80d609..dbaa604 100644
--- a/Rakefile
+++ b/Rakefile
@@ -55,6 +56,30 @@ namespace :benchmark do
                             args.repeats)
   end
 
+  desc 'Allocation counts'
+  task allocations: :support do
+    if RUBY_VERSION < '2.1'
+      $stderr.puts "Cannot count allocations on #{RUBY_VERSION}."
+      exit 1
+    end
+
+    begin
+      require 'allocation_tracer'
+    rescue LoadError
+      $stderr.puts "Allocation tracking requires the gem 'allocation_tracer'."
+      exit 1
+    end
+
+    ObjectSpace::AllocationTracer.trace do
+      require 'mime/types'
+    end
+
+    count = ObjectSpace::AllocationTracer.allocated_count_table.values.
+      inject(:+)
+
+    puts "Allocations: #{count}"
+  end
+
   desc 'Show object counts'
   task objects: :support do
     GC.start
```

If you also want to add a benchmark for the hotspot finder, that would be awesome. This way, I can also use these to ensure no regressions.

**Comment by halostatue** [comment]
(I check for the `allocation_tracer` gem because mime-types still supports Ruby < 2.1 and I can’t even put it in the Rakefile/gemspec without breaking Travis.)

**Comment by schneems** [comment]
I tried adding this:

```ruby
desc 'Allocation counts'
task allocations: :support do
  require 'pp'

  if RUBY_VERSION < '2.1'
    $stderr.puts "Cannot count allocations on #{RUBY_VERSION}."
    exit 1
  end

  begin
    require 'allocation_tracer'
  rescue LoadError
    $stderr.puts "Allocation tracking requires the gem 'allocation_tracer'."
    exit 1
  end

  r = ObjectSpace::AllocationTracer.trace do
    require 'mime/types'
  end

  count = ObjectSpace::AllocationTracer.allocated_count_table.values.
    inject(:+)

  pp r.sort {|x, y| x.last.first <=> y.last.first }.reverse.first(100)
  puts "Allocations: #{count}"
end
```

But I always get an error that allocation tracer can't be found since we're modifying the load path in the `:support` task. I'm jumping on a plane in a few minutes.

**Comment by schneems** [comment]
Maybe it's not `:support` but i'm not sure. It's failing but i've definitely got it on my system.

**Comment by halostatue** [comment]
Hmm. Let me try adding it locally. If it works, we’ll just merge this and I’ll add the benchmark later.

**Comment by schneems** [comment]
I pushed the benchmarks, got it working. I think in my haste i was running it with `bundle exec` but of course it's not in the Gemfile.lock

**Comment by halostatue** [comment]
Thanks. I ran out of time to look at this over the last couple of days. I’m merging this and will release a new version in the next few days.

**Pull request merged**